File: src/ladym/storage/vector_index.py
"""Vector index abstraction with two interchangeable implementations.

:class:`InMemoryVectorIndex` is a pure-Python numpy index used in tests and tiny workspaces.
:class:`SQLiteVecIndex` wraps the ``sqlite-vec`` loadable extension for production-scale
persistent storage. Both implement the same :class:`VectorIndex` ABC.
"""

from __future__ import annotations

from abc import ABC, abstractmethod

import numpy as np


class VectorIndex(ABC):
    """Insert/query/delete vectors keyed by an arbitrary id."""

    dim: int

    @abstractmethod
    def upsert(self, item_id: str, vector: list[float]) -> None: ...

    @abstractmethod
    def search(self, query: list[float], top_k: int = 10) -> list[tuple[str, float]]:
        """Return ``(id, similarity)`` pairs, most similar first."""

    @abstractmethod
    def delete(self, item_id: str) -> None: ...

    @abstractmethod
    def __len__(self) -> int: ...


class InMemoryVectorIndex(VectorIndex):
    """Brute-force numpy cosine index. Deterministic, perfect for tests."""

    def __init__(self, dim: int):
        self.dim = dim
        self._ids: list[str] = []
        self._id_to_idx: dict[str, int] = {}
        self._mat: np.ndarray | None = None  # shape (N, dim)

    def upsert(self, item_id: str, vector: list[float]) -> None:
        vec = np.asarray(vector, dtype=np.float32)
        if vec.shape[0] != self.dim:
            raise ValueError(f"vector dim {vec.shape[0]} != index dim {self.dim}")
        # L2-normalise so cosine == dot product
        n = np.linalg.norm(vec)
        if n > 0:
            vec = vec / n
        if self._mat is None:
            self._mat = np.zeros((1, self.dim), dtype=np.float32)
            self._mat[0] = vec
            self._ids.append(item_id)
            self._id_to_idx[item_id] = 0
            return
        if item_id in self._id_to_idx:
            self._mat[self._id_to_idx[item_id]] = vec
        else:
            self._mat = np.vstack([self._mat, vec[None, :]])
            idx = len(self._ids)
            self._ids.append(item_id)
            self._id_to_idx[item_id] = idx

    def search(self, query: list[float], top_k: int = 10) -> list[tuple[str, float]]:
        if self._mat is None or len(self._ids) == 0:
            return []
        q = np.asarray(query, dtype=np.float32)
        n = np.linalg.norm(q)
        if n > 0:
            q = q / n
        sims = self._mat @ q  # both normalised → dot == cosine
        k = min(top_k, len(self._ids))
        # argpartition then sort the top-k slice
        idx = np.argpartition(-sims, k - 1)[:k]
        idx = idx[np.argsort(-sims[idx])]
        return [(self._ids[i], float(sims[i])) for i in idx]

    def delete(self, item_id: str) -> None:
        if item_id not in self._id_to_idx or self._mat is None:
            return
        idx = self._id_to_idx.pop(item_id)
        last = len(self._ids) - 1
        # swap with last, then trim
        if idx != last:
            self._mat[idx] = self._mat[last]
            swapped_id = self._ids[last]
            self._ids[idx] = swapped_id
            self._id_to_idx[swapped_id] = idx
        self._ids.pop()
        self._mat = self._mat[:last] if last > 0 else None

    def __len__(self) -> int:
        return len(self._ids)


class SQLiteVecIndex(VectorIndex):
    """Persistent index backed by the ``sqlite-vec`` loadable extension.

    Each LadyM store keeps one ``vec_memories`` virtual table. We delegate to sqlite-vec for
    ANN search; sqlite-vec uses cosine distance natively when vectors are normalised.
    """

    def __init__(self, conn, dim: int):  # type: ignore[no-untyped-def]
        self.conn = conn
        self.dim = dim
        try:
            import sqlite_vec  # type: ignore

            conn.enable_load_extension(True)
            conn.load_extension(sqlite_vec.loadable_path())
            conn.enable_load_extension(False)
        except Exception as e:  # pragma: no cover - environment-specific
            raise RuntimeError(
                "Failed to load sqlite-vec. Install with: pip install sqlite-vec"
            ) from e
        conn.execute(
            f"CREATE VIRTUAL TABLE IF NOT EXISTS vec_memories USING vec0("
            f"id TEXT PRIMARY KEY, embedding float[{dim}] distance_metric=cosine)"
        )
        conn.commit()

    def upsert(self, item_id: str, vector: list[float]) -> None:
        blob = np.asarray(vector, dtype=np.float32).tobytes()
        # vec0 doesn't support upsert directly; delete-then-insert.
        self.conn.execute("DELETE FROM vec_memories WHERE id = ?", (item_id,))
        self.conn.execute(
            "INSERT INTO vec_memories(id, embedding) VALUES (?, ?)",
            (item_id, blob),
        )
        self.conn.commit()

    def search(self, query: list[float], top_k: int = 10) -> list[tuple[str, float]]:
        blob = np.asarray(query, dtype=np.float32).tobytes()
        rows = self.conn.execute(
            "SELECT id, distance FROM vec_memories "
            "WHERE embedding MATCH ? AND k = ? "
            "ORDER BY distance",
            (blob, top_k),
        ).fetchall()
        # sqlite-vec returns cosine *distance* (= 1 - similarity for normalised vectors)
        return [(r[0], max(0.0, 1.0 - r[1])) for r in rows]

    def delete(self, item_id: str) -> None:
        self.conn.execute("DELETE FROM vec_memories WHERE id = ?", (item_id,))
        self.conn.commit()

    def __len__(self) -> int:
        row = self.conn.execute("SELECT COUNT(*) FROM vec_memories").fetchone()
        return int(row[0]) if row else 0


def make_index(prefer_sqlite: bool, conn, dim: int) -> VectorIndex:  # type: ignore[no-untyped-def]
    """Factory: try sqlite-vec, fall back to in-memory (so tests never hard-fail)."""
    if prefer_sqlite:
        try:
            return SQLiteVecIndex(conn, dim)
        except Exception:
            pass
    return InMemoryVectorIndex(dim)

File: storage/vector_index.go
package storage

import (
	"fmt"
	"math"
	"sort"
)

// SearchHit is one (id, similarity) pair returned by Search.
type SearchHit struct {
	ID         string
	Similarity float64
}

// VectorIndex inserts/queries/deletes vectors keyed by an arbitrary id.
type VectorIndex interface {
	Upsert(itemID string, vector []float32) error
	Search(query []float32, topK int) []SearchHit
	Delete(itemID string)
	Len() int
}

// InMemoryVectorIndex is a brute-force cosine index (deterministic, perfect for
// tests and small workspaces). It replaces the sqlite-vec extension used by the
// Python port with a pure-Go equivalent; vectors are still persisted as BLOBs
// in the store so a reopened store can rebuild the index.
type InMemoryVectorIndex struct {
	dim     int
	ids     []string
	idToIdx map[string]int
	mat     [][]float32 // shape (N, dim), L2-normalised
}

// NewInMemoryVectorIndex returns an empty index of the given dim.
func NewInMemoryVectorIndex(dim int) *InMemoryVectorIndex {
	return &InMemoryVectorIndex{dim: dim, idToIdx: map[string]int{}}
}

func l2Norm(vec []float32) float64 {
	var n float64
	for _, v := range vec {
		n += float64(v) * float64(v)
	}
	return math.Sqrt(n)
}

func (ix *InMemoryVectorIndex) Upsert(itemID string, vector []float32) error {
	if len(vector) != ix.dim {
		return fmt.Errorf("vector dim %d != index dim %d", len(vector), ix.dim)
	}
	vec := append([]float32{}, vector...)
	if n := l2Norm(vec); n > 0 {
		for i := range vec {
			vec[i] = float32(float64(vec[i]) / n)
		}
	}
	if idx, ok := ix.idToIdx[itemID]; ok {
		ix.mat[idx] = vec
		return nil
	}
	ix.mat = append(ix.mat, vec)
	ix.idToIdx[itemID] = len(ix.ids)
	ix.ids = append(ix.ids, itemID)
	return nil
}

func (ix *InMemoryVectorIndex) Search(query []float32, topK int) []SearchHit {
	if len(ix.ids) == 0 {
		return nil
	}
	q := append([]float32{}, query...)
	if n := l2Norm(q); n > 0 {
		for i := range q {
			q[i] = float32(float64(q[i]) / n)
		}
	}
	type scored struct {
		id  string
		sim float64
	}
	all := make([]scored, 0, len(ix.ids))
	for i, vec := range ix.mat {
		var dot float64
		for j := range vec {
			dot += float64(vec[j]) * float64(q[j])
		}
		all = append(all, scored{ix.ids[i], dot})
	}
	sort.SliceStable(all, func(a, b int) bool {
		if all[a].sim != all[b].sim {
			return all[a].sim > all[b].sim
		}
		return all[a].id < all[b].id
	})
	k := topK
	if k > len(all) {
		k = len(all)
	}
	out := make([]SearchHit, 0, k)
	for _, s := range all[:k] {
		out = append(out, SearchHit{ID: s.id, Similarity: s.sim})
	}
	return out
}

func (ix *InMemoryVectorIndex) Delete(itemID string) {
	idx, ok := ix.idToIdx[itemID]
	if !ok {
		return
	}
	delete(ix.idToIdx, itemID)
	last := len(ix.ids) - 1
	if idx != last {
		ix.mat[idx] = ix.mat[last]
		swappedID := ix.ids[last]
		ix.ids[idx] = swappedID
		ix.idToIdx[swappedID] = idx
	}
	ix.ids = ix.ids[:last]
	ix.mat = ix.mat[:last]
}

func (ix *InMemoryVectorIndex) Len() int { return len(ix.ids) }
